### PR TITLE
Fully lock adding node queues during hinted handoff

### PR DESCRIPTION
**Pull request by otoolep**
I believe this change address the issues with hinted-handoff not fully replicating all data to nodes that come back online after an outage. A detailed explanation follows.

During testing of hinted-handoff (HH) under various scenarios, HH stats showed that the HH Processor was occasionally encountering errors while unmarshalling hinted data. This error was not handled completely correctly, and in clusters with more than 3 nodes, this could cause the HH service to stall until the node was restarted. This was the high-level reason why HH data was not being replicated.

Furthermore by watching, at the byte-level, the hinted-handoff data it could be seen that HH segment block lengths were getting randomly set to 0, but the block data itself was fine (Block data contains hinted writes). This was the root cause of the unmarshalling errors outlined above. This, in turn, was tracked down to the HH system opening each segment file multiple times concurrently. This is not thread-safe, so these mutiple `open` calls were corrupting the file.

Finally, the reason a segment file was being opened multiple times in parallel was because `WriteShard` on the HH Processor was manipulating HH node queues in an unsafe manner. Since `WriteShard` can be called concurrently this was adding queues for the same node more than once, and each queue-addition results in opening segment files.

This change fixes the locking in `WriteShard` such the check for an existing HH queue for a given node is performed in a synchronized manner.

Before this change I could cause HH to fail 100% within a couple of minutes due to corrupt HH data, under specific testing. With this change in place everything looks fine.

### CODE REVIEW

File: services/hh/processor.go [hunk 1]
```diff
@@ -128,11 +127,27 @@ func (p *Processor) addQueue(nodeID uint64) (*queue, error) {
...
 func (p *Processor) WriteShard(shardID, ownerID uint64, points []models.Point) error {
+	p.mu.RLock()
 	queue, ok := p.queues[ownerID]
+	p.mu.RUnlock()
 	if !ok {
-		var err error
-		if queue, err = p.addQueue(ownerID); err != nil {
+		if err := func() error {
+			// Check again under write-lock.
+			p.mu.Lock()
```
Comment (otoolep) [inline services/hh/processor.go]: This locking pattern is the same as used by the `expvar` package in the standard libary.

**Comment by otoolep** [comment]
@jwilder

File: services/hh/processor.go [hunk 2]
```diff
@@ -128,11 +127,27 @@ func (p *Processor) addQueue(nodeID uint64) (*queue, error) {
 	return queue, nil
 }
 
+// WriteShard writes hinted-handoff data for the given shard and node. Since it may manipulate
+// hinted-handoff queues, and be called concurrently, it takes a lock during queue access.
 func (p *Processor) WriteShard(shardID, ownerID uint64, points []models.Point) error {
+	p.mu.RLock()
 	queue, ok := p.queues[ownerID]
```
Comment (otoolep) [inline services/hh/processor.go]: This is root cause. `p.queues` was not being locked when checked for an entry for `ownerID`.

**Comment by corylanou** [comment]
Would be nice it we were able to validate this via a test.

**Comment by jwilder** [comment]
:+1: That definitely needs synchronization.

**Comment by otoolep** [comment]
@corylanou - I am open to suggestions but I not sure how this is any
different from any piece of code in our system, which if the locking is
wrong, we go in and fix the lock. It's difficult to create a unit test for
this since it may always pass since it's a race.

One way we now can check - and I will make sure we do - is to always run
one of the burn-in boxes with race detection on. That way we will start
catching issues like this during system testing.

**Comment by otoolep** [comment]
@corylanou - let me know if you have any ideas for testing this. I will be enabling race-detection on a single burn-in now.

**Comment by corylanou** [comment]
@otoolep yeah, I don't have an easy answer.  We could possibly add a test that does a bunch of routines all calling HH and `hope` the race detector catches it, but again, it's hard to write a test that can reproduce this consistently, hence why the race detector exists.  I think your idea of the burn in box with race enabled is a great step in the right direction.

**Pull request merged**